### ISSUE TITLE
DPL-362 As GSU (Alan K) I want to allocate a unique COG-UK ID to all historic Heron samples so that we can use this instead of Root Sample ID within our internal tracking.

### ISSUE THREAD

**Issue by TWJW-SANGER** [sanger/lighthouse#576]
**User story**
As PAM (Alan K) I want to allocate a unique COG-UK ID to all historic Heron samples so that we can use this instead of Root Sample ID within our internal tracking.

**Who are the primary contacts for this story**
Alan K
Cristina A
Carl R

**Acceptance criteria**
To be considered successful the solution must allow:
- [x] All samples within the MLWH table lighthouse_sample will be allocated a cog_uk_id
- [x] The cog_uk_id's used must be unique
- [x] New samples being registered in the lighthouse_sample  table after this modification - must also be given unique COG-UK-ids
- [x] Attempts to insert samples with duplicates COG-UK-IDs into the lighthouse_sample  table should fail gracefully

**Dependencies**
This story is blocked by the following dependencies:
- #509
It only makes sense to do this once all new incoming samples from the lighthouse labs received COG-UK-ids

**Additional context**
Current reporting systems rely on the the Root Sample ID provided by the Lighthouses, which is copied around various systems - however this is manually generated by lighthouses and unreliable.
By moving towards using our own id we can gradually reduce the number of systems that have to be patched everytime the Root Sample ID is updated/corrected.

**Comment by sdjmchattie** [comment]
After some preliminary investigation, the following is true of the data today:

- There are ~51.5 million lighthouse_sample table rows.
  - ~48.5 million don't have a COG UK ID and these will need to be assigned one.
  - ~3 million already have a COG UK ID.
- There are about 8.3 million samples in the sample table.
  - ~1.3 million of these have no COG UK ID.
  - The remaining 7 million do have a COG UK ID.
- Comparing samples across the tables, joining on the UUID of the sample, the following discrepancies can be identified.
  - 363,766 have only got a COG UK ID in the sample table -- possibly because of manifest insertion into sequence scape as well as being sent via SFTP.
  - 11 have only got a COG UK ID in the lighthouse_sample table -- the cause of which is currently unknown and more investigation is needed.
  - 1,456 have a COG UK ID in both tables, but it is different in each table.  It's possible that this is a manifestation of the underlying cause where the data was fixed by https://ssg-confluence.internal.sanger.ac.uk/display/PSDPUB/Duplicate+COG+UK+IDs+May+2021 particularly since the bad samples are all after June 2021 when a fix was previously applied.

The bad samples have been exported into Excel files, but won't be attached here on GitHub.

**Comment by sdjmchattie** [comment]
Quick update on the 11 where the COG UK ID is only in the lighthouse_sample table.  They do show the correct COG UK IDs in the Sequencescape webpage for the samples.  It seems those were the subject of a data tweak in Sequencescape where Neil has commented that:

> Sample xxxxxxx moved from 6187 to 6198 requested via RTDPL-020 using sample_move.rb

Some of the samples on that plate have the COG UK IDs in the sample table so I expect the answer here is to `touch` the samples in the Sequencescape Ruby on Rails console and have it update the information in the MLWH.  Then these should be consistent.

**Comment by sdjmchattie** [comment]
The 1,456 samples are the ones that are of most concern because we don't yet know which COG UK ID is the one that has been used to process samples.  I am going through the database tables, the picking events and the plates in Sequencescape to identify what has happened for each of the 25 plates affected.  I'm about halfway done with these and have a pretty common theme through them all with some minor variations.  I am compiling a document explaining each and it will contain suggestions for how we can fix all 1,456 samples and what we can do to avoid this situation in future.

**Comment by sdjmchattie** [comment]
This is an except from the full summary and proposed solution section of the data patch repository:

> In cases where COG UK IDs can be found in one table and not the other, we should copy the value to the table where it is missing.  In the case of updating the `sample` table, this will be achieved by touching the data in Sequencescape itself.  It is worth noting, that this will generate duplicates in the COG UK ID column which was not what we desired according to the criteria of this data patch.
> 
> Where the COG UK IDs are inconsistently recorded between the two tables, we should prefer the value in the `sample` table which represents the original COG UK ID a sample was given before being recorded in Sequencescape and is the value used to track the sequencing of that sample.
> 
> There are duplicated COG UK IDs in the `lighthouse_sample` table already which affect only 79 COG UK IDs.  All of these are from Milton Keynes and where they are duplicated they appear to refer to a second well of the same sample.  This might be due to a previous data patch, but this is not clear.  It's also not clear whether half of these are actually some sort of test data.  Maybe the date ranges, plate barcodes or root sample IDs will reveal this.
> 
> There are 4,721 COG UK IDs to be imported from the set of 373,766 missing in the `lighthouse_sample` table which would cause duplicate usage of COG UK IDs.  Closer inspection of some of these indicates that the COG UK IDs were used in both April 2020 and then again in October 2021.  Those used in April 2020 were uploaded via manifest.  These do appear to be different samples and may have been caused by inconsistent generation of COG UK IDs in the early days of the project.
> 
> As many of the samples that have duplicate COG UK IDs have already passed through Sequencescape and appear to have been sequenced, it would be very difficult or impossible to change their COG UK IDs now.  These duplicates are going to have to be accepted as they are now, which doesn't help with the anticipated use of a unique COG UK ID for internal tracking.  Unique COG UK IDs in future could be assured, but these samples are already too far through the process to change.
> 
> Another possibility would be that a new column be added to the `lighthouse_sample` table that is used for internal tracking.  This would duplicate the value of the COG UK ID for almost all samples.  Those where we know uniqueness is not assured, we could add suffixes to the duplicates so that they appear with `-1`, `-2` after the COG UK ID, for example.
> 
> Once a satisfactory strategy has been decided, the rest of the unpopulated COG UK IDs are simply as yet unassigned.  We can then run a script as mentioned in index 2 of the [Approach](#approach) to fill in every COG UK ID throughout the table.  The script would, plate-by-plate, request new COG UK IDs from Baracoda for the missing ones, and then write them all to the samples before repeating for the next plate.  If this needed to be run again in future, it would therefore only need to apply changes to the samples where a COG UK ID is missing.

The rest of this document and associated data exploration scripts and outputs can be accessed via the private GitLab instance https://gitlab.internal.sanger.ac.uk/psd/data-patch-archive/-/tree/main/patches/2022/0815-dpl-362-populate-cog-uk-ids-in-lighthouse-sample-table

**Comment by sdjmchattie** [comment]
I've emailed Alan with details of our findings and suggestions.  For now, I'm marking it as On Hold and we can come back to it in a few days after Alan has had a chance to read and we can have a meeting to decide how, or even if, to move forwards.

**Comment by sdjmchattie** [comment]
Further investigation has been done following a meeting last Friday to discuss ways forward.  There are a number of samples that would cause duplication of COG UK IDs in the `lighthouse_sample` table if copied without consideration.  There are reasons why these duplicates might exist but more clarity has been given to how those duplicates could be mitigated.  Either we need to change the COG UK IDs, but that is fraught with danger of errors and future confusion, or we need to allow duplicate COG UK ID usage before a certain date.  The latter is my suggestion.  We can enforce future uniqueness by using a flagging column and creating a unique index over COG UK ID and this new flag column.  Historical values will be given `NULL` as a value in the flag column, and all future values will use a default known value such as `1`.  We are awaiting feedback from GSU about this to ensure they are happy with the approach before we adopt it.

**Comment by sdjmchattie** [comment]
The scripts and SQL commands required to make this happen have been completed and are now in review.  The code is on GitLab due to the sensitive data it is modifying.  Therefore the pull request is at https://gitlab.internal.sanger.ac.uk/psd/data-patch-archive/-/merge_requests/3

The README there indicates the exploration of the data, the suggested course of action, and SQL commands / Python scripts that will apply the fixes.  The population of COG UK IDs that are missing is quite intense and there are a lot to fill in so we might want to run that on a Sanger server with access to MLWH.  `tmux` could be used to allow it to run unattended as I expect it to take a few hours to complete.

**Comment by sdjmchattie** [comment]
The script to apply COG UK IDs to samples that are missing one has been run on production two evenings now.  Once on the 13th September for 5 hours, where 17.9 million samples were updated, and again on 15th September for 5 hours where 23.5 million samples were updated.  The performance improvement was due to changes that the database admins advised after the first run caused some replication issues that have now been resolved.

There are 6.3 million samples still to update and four SQL statements needed to complete the update by altering the `lighthouse_sample` table to prevent NULL and duplicate COG UK IDs being inserted in future.  Due to the weekend and the bank holiday in the UK, these final updates will be performed next week on Tuesday and Wednesday, so that the DBA team will be available the day after updates in case of further replication issues.

**Comment by sdjmchattie** [comment]
6,333,213 remaining samples were updated with COG UK IDs yesterday evening.  This only needed 57 minutes to complete.  The table schema still needs updating to prevent duplicate COG UK IDs in the future, but the data is now up to date.

There are 4,799 duplicated COG UK IDs from the historic data in the table.  There shouldn't be any more in the future once the schema has been updated.

**Comment by TWJW-SANGER** [comment]
DB Schema updated successfully.
It is no longer possible to add lighthouse samples to the table without a unique COG UK ID.
